scipy/linalg/_decomp_cossin.py

```diff
@@ -92,73 +83,63 @@ def cossin(X, partitioning=None, compute_u=True, compute_vh=True):
-        raise ValueError("invalid submatrix dimensions {} to build a square "
-                         "matrix, maybe you forgot to use \"partitioning\" keyword?".format(shapes))
-    return _cossin(x11, x12, x21, x22, compute_u, compute_vh)
-
 
-def _cossin(x11, x12, x21, x22, compute_u, compute_vh):
-    p, q = np.shape(x11)
-    m = np.shape(x22)[0] + p
-
-    x11, x12, x21, x22 = _cast_all_to_largest(x11, x12, x21, x22)
-
-    iscomp = np.iscomplexobj(x11)
-    func_name = "uncsd" if iscomp else "orcsd"
-
-    csd, csd_lwork = get_lapack_funcs([func_name, func_name + "_lwork"],
-                                      [x11])
-    lwork = _compute_lwork(csd_lwork, m=m, p=p, q=q)
-
-    lwork_args = (dict(lwork=lwork[0], lrwork=lwork[1]) if iscomp else
-                  dict(lwork=lwork))
-
-    csd_args = dict(x11=x11,
-                    x12=x12,
-                    x21=x21,
-                    x22=x22,
-                    compute_u1=compute_u,
-                    compute_u2=compute_u,
-                    compute_v1t=compute_vh,
-                    compute_v2t=compute_vh,
-                    trans=False,
-                    signs=False)
-
-    csd_args.update(lwork_args)
-
-    x11o, x12o, x21o, x22o, theta, u1, u2, v1h, v2h, info = csd(
-        **csd_args)
-
-    method_name = csd.typecode + func_name
+        x11, x12, x21, x22 = X[:p, :q], X[:p, q:], X[p:, :q], X[p:, q:]
+        mmp, mmq = X[p:, q:].shape
+        m = X.shape[0]
+    elif not isinstance(X, Iterable):
+        raise ValueError("When p and q are None, X must be an Iterable"
+                         " containing the subblocks of X")
+    else:
+        if len(X) != 4:
+            raise ValueError("When p and q are None, exactly four arrays"
+                             " should be in X, got {}".format(len(X)))
+
+        x11, x12, x21, x22 = [np.atleast_2d(x) for x in X]
+        p, q = x11.shape
+        mmp, mmq = x22.shape
+
+        if x12.shape != (p, mmq):
+            raise ValueError("Invalid x12 dimensions: desired {}, "
+                             "got {}".format((p, mmq), x12.shape))
+
+        if x21.shape != (mmp, q):
+            raise ValueError("Invalid x21 dimensions: desired {}, "
+                             "got {}".format((mmp, q), x21.shape))
+
+        if p+mmp != q+mmq:
+            raise ValueError("The subblocks have compatible sizes but "
+                             "don't form a square array (instead they form a"
+                             " {}x{} array). This might be due to missing "
+                             "p, q arguments.".format(p+mmp, q+mmq))
+
+        m = p + mmp
+
+    cplx = any([np.iscomplexobj(x) for x in [x11, x12, x21, x22]])
+    driver = "uncsd" if cplx else "orcsd"
+    csd, csd_lwork = get_lapack_funcs([driver, driver + "_lwork"],
+                                      [x11, x12, x21, x22])
+
+    lwork = _compute_lwork(csd_lwork, m=p+q, p=p, q=q)
```

Comment:
`m=p+q` is incorrect

Reply:
> Ah, it is an error handling problem that slips through with one of the arguments:
> 
> ```
> python -c "from scipy.linalg import cossin;from scipy.stats import unitary_group; cossin(unitary_group.rvs(4), 1, 2)"
>  ** On entry to ZUNCSD parameter number 22 had an illegal value
> ```

@ilayn - this worries me that this error slipped through - shouldn't this be handled by the get_lapack_funcs / lapack wrappers and return a nice error status with `info=-22`? Instead it just crashed seemingly!

Reply:
yeah, I confirmed that starting with `python runtests.py --ipython` behaves differently on Linux than on Windows.
By setting  L130 (incorrectly) to    `lwork = _compute_lwork(csd_lwork, m=3, p=p, q=q)` and  in ipython: 

```
 from scipy.linalg import cossin; from scipy.stats import unitary_group; cossin(unitary_group.rvs(4), 1, 2) 
```

On MacOSX: (4, 1, 2) passes!! However `cossin(unitary_group.rvs(4), 2, 2)` fails with 
```
ry_group.rvs(4), 2, 2) 

~/dev/proj/scipy/build/testenv/lib/python3.7/site-packages/scipy/linalg/_decomp_cossin.py in cossin(X, p, q, minus_upper, compute_u, compute_vh)
    159     if info < 0:
    160         raise ValueError('illegal value in argument {} of internal {}'
--> 161                          .format(-info, method_name))
    162     if info > 0:
    163         raise LinAlgError("{} did not converge: {}".format(method_name, info))

ValueError: illegal value in argument 24 of internal zuncsd
``` which is good I suppose. 

On Linux (in a docker container): 

```
 ** On entry to ZUNCSD parameter number 22 had an illegal value
```
and the ipython process crashes!

Reply:
I guess the difference in parameter number is due to difference in how ILAENV returns different sizes for parameters on different OSes. And for some reason the lapack errors crash scipy on linux and these crashes can't be caught. So...I am guessing that we can't really do much about this.